h/h_api/bulk_api/id_references.py

```diff
@@ -0,0 +1,81 @@
+"""Objects for referencing and de-referencing ids."""
+
+from collections import defaultdict
+
+from h.h_api.enums import DataType
+from h.h_api.exceptions import UnpopulatedReferenceError
+
+
+class IdReferences:
+    """
+    A store of id references which can discover and populate them.
+
+    If we create one object which relies on another (group membership on a
+    group) we will not know the id of the group until the group has been
+    updated.
+
+    To get around this you can add a temporary id reference to the group,
+    e.g. "my_ref" which is swapped for a concrete id when the group is
+    created. We can reference this in the group membership with
+    `{"$ref": "my_id"}`.
+
+    It's this classes' job to store and fill out these references.
+    """
+
+    REF_KEY = "$ref"
+
+    def __init__(self):
+        self.id_refs = defaultdict(dict)
```

Comment:
`self.concrete_ids` might be a better name. They're keyed by id refs, but the point of this dict seems to be to store and retrieve the concrete ids.

Reply:
Should this have a leading underscore?

Reply:
I've changed this to `_ref_to_concrete`, to let you know what we're mapping from and to.